src/action-providers.ts

```diff
@@ -21,18 +21,87 @@ class RefactoringActionProvider implements vscode.CodeActionProvider {
     const ast = t.parse(document.getText());
     const selection = createSelectionFromVSCode(range);
 
-    return this.refactorings
-      .filter(refactoring => this.canPerform(refactoring, ast, selection))
-      .map(refactoring => this.buildCodeActionFor(refactoring));
+    const applicableRefactorings: RefactoringWithActionProvider[] = [];
+
+    const onCanPeform = (
+      path: t.NodePath<any>,
+      refactoring: RefactoringWithActionProvider
+    ) => {
+      if (isLegacyActionProvider(refactoring.actionProvider)) {
+        return;
+      }
+
+      if (refactoring.actionProvider.updateMessage) {
+        refactoring.actionProvider.updateMessage(path);
+      }
+
+      applicableRefactorings.push(refactoring);
+    };
+
+    t.traverseAST(ast, {
+      enter: (path: t.NodePath<any>) => {
+        this.refactorings.forEach(refactoring =>
+          this.canPerform(refactoring, path, selection, onCanPeform)
+        );
+      }
+    });
+
+    const applicableLegacyRefactorings = this.refactorings.filter(refactoring =>
+      this.canPerformLegacy(refactoring, ast, selection)
+    );
+
+    return [...applicableRefactorings, ...applicableLegacyRefactorings].map(
+      refactoring => this.buildCodeActionFor(refactoring)
+    );
   }
 
   private canPerform(
+    refactoring: RefactoringWithActionProvider,
+    path: t.NodePath<any>,
+    selection: Selection,
+    onCanPerform: (
+      matchedPath: t.NodePath<any>,
+      refactoring: RefactoringWithActionProvider
+    ) => void
+  ) {
+    if (isLegacyActionProvider(refactoring.actionProvider)) {
+      return;
+    }
+
+    const visitor: t.Visitor = refactoring.actionProvider.createVisitor(
+      selection,
+      path => onCanPerform(path, refactoring),
+      refactoring
+    );
+
+    this.visit(visitor, path);
+  }
+
+  private visit(visitor: any, path: t.NodePath<any>) {
+    const node: t.Node = path.node;
+
+    try {
+      if (typeof visitor[node.type] === "function") {
+        visitor[node.type](path);
+      } else if (typeof visitor[node.type] === "object") {
+        visitor[node.type].enter(path);
+      }
+    } catch (_) {
+      // Silently fail, we don't care why it failed (e.g. code can't be parsed).
+    }
+  }
+
+  private canPerformLegacy(
     refactoring: RefactoringWithActionProvider,
     ast: t.AST,
     selection: Selection
   ) {
     try {
-      return refactoring.actionProvider.canPerform(ast, selection);
+      return (
+        isLegacyActionProvider(refactoring.actionProvider) &&
+        typeof refactoring.actionProvider.canPerform === "function" &&
```

Comment:
Is line 102 necessary? Isn't it covered by `isLegacyActionProvider()` already?

```suggestion
        isLegacyActionProvider(refactoring.actionProvider) &&
```

Reply:
Alternative suggestion: change the type of `refactoring` to only be a Legacy one + use the guard clause to filter out the non-Legacy ones before calling the function. That will preserve the original behavior of the function and split the responsibilities.